src/flamenco/vm/fd_vm_private.h

```diff
@@ -309,25 +309,35 @@ fd_vm_mem_haddr( fd_vm_t const *    vm,
                  uchar              write,           /* 1 if the access is a write, 0 if it is a read */
                  ulong              sentinel,
                  uchar *            is_multi_region ) {
-  ulong vaddr_hi  = vaddr >> 32;
-  ulong region    = fd_ulong_min( vaddr_hi, 5UL );
-  ulong offset    = vaddr & 0xffffffffUL;
+  ulong vaddr_hi       = vaddr >> 32;
+  ulong region         = fd_ulong_min( vaddr_hi, 5UL );
+  ulong offset         = vaddr & 0xffffffffUL;
+  uchar direct_mapping = FD_FEATURE_ACTIVE( vm->instr_ctx->slot_ctx, bpf_account_data_direct_mapping );
 
   /* Stack memory regions have 4kB unmapped "gaps" in-between each frame.
     https://github.com/solana-labs/rbpf/blob/b503a1867a9cfa13f93b4d99679a17fe219831de/src/memory_region.rs#L141
     */
-  if ( FD_UNLIKELY( region == 2UL ) ) {
+  if( FD_UNLIKELY( region == 2UL ) ) {
+    /* Stack frame gaps are only enabled when direct mapping is disabled. vm_gap_shift is 63 when direct mapping is enabled,
+       and 12 otherwise.
+       https://github.com/anza-xyz/agave/blob/v2.0.10/programs/bpf_loader/src/syscalls/mod.rs#L297 */
+    uchar vm_gap_shift = direct_mapping ? 63 : 12;
+
     /* If an access starts in a gap region, that is an access violation */
-    if ( !!( vaddr & 0x1000 ) ) {
+    uchar is_in_gap = ( offset >> vm_gap_shift ) & 1;
+
+    /* https://github.com/solana-labs/rbpf/blob/v0.8.5/src/memory_region.rs#L140-L145
+       Region OOB access checks are performed later in our implementation. */
+    if( FD_UNLIKELY( is_in_gap ) ) {
       return sentinel;
     }
 
     /* To account for the fact that we have gaps in the virtual address space but not in the 
        physical address space, we need to subtract from the offset the size of all the virtual
        gap frames underneath it.
        
-       https://github.com/solana-labs/rbpf/blob/b503a1867a9cfa13f93b4d99679a17fe219831de/src/memory_region.rs#L147-L149 */
-    ulong gap_mask = 0xFFFFFFFFFFFFF000;
+       https://github.com/solana-labs/rbpf/blob/v0.8.5/src/memory_region.rs#L137 */
+    ulong gap_mask = ( 0xFFFFFFFFFFFFFFFFUL << vm_gap_shift );
     offset = ( ( offset & gap_mask ) >> 1 ) | ( offset & ~gap_mask ); 
   }
```

Comment:
This whole block is only needed if the gaps are enabled, so you can keep the inside as-is and just add the direct mapping condition to the if statement.

Reply:
I think it's still necessary - if direct mapping is enabled but the access is not in a stack frame gap, the offset would be calculated incorrectly.

Reply:
the offset adjustment is just to account for the fact we have virtual stack gaps but not physical ones - if there are no virtual gaps no adjustment _should_ be needed